sources/directconnect/connection.go

```diff
@@ -0,0 +1,139 @@
+package directconnect
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/directconnect"
+	"github.com/overmindtech/aws-source/sources"
+	"github.com/overmindtech/sdp-go"
+)
+
+func connectionOutputMapper(_ context.Context, _ *directconnect.Client, scope string, _ *directconnect.DescribeConnectionsInput, output *directconnect.DescribeConnectionsOutput) ([]*sdp.Item, error) {
+	items := make([]*sdp.Item, 0)
+
+	for _, connection := range output.Connections {
+		attributes, err := sources.ToAttributesCase(connection, "tags")
+		if err != nil {
+			return nil, &sdp.QueryError{
+				ErrorType:   sdp.QueryError_OTHER,
+				ErrorString: err.Error(),
+				Scope:       scope,
+			}
+		}
+
+		item := sdp.Item{
+			Type:            "directconnect-connection",
+			UniqueAttribute: "connectionId",
+			Attributes:      attributes,
+			Scope:           scope,
+			Tags:            tagsToMap(connection.Tags),
+		}
+
+		if connection.LagId != nil {
+			// +overmind:link directconnect-lag
+			item.LinkedItemQueries = append(item.LinkedItemQueries, &sdp.LinkedItemQuery{
+				Query: &sdp.Query{
+					Type:   "directconnect-lag",
+					Method: sdp.QueryMethod_GET,
+					Query:  *connection.LagId,
+					Scope:  scope,
+				},
+				BlastPropagation: &sdp.BlastPropagation{
+					// Changes to the lag will affect this
+					In: true,
+					// We can't affect the lag
+					Out: false,
+				},
+			})
+		}
+
+		if connection.Location != nil {
+			// +overmind:link directconnect-location
+			item.LinkedItemQueries = append(item.LinkedItemQueries, &sdp.LinkedItemQuery{
+				Query: &sdp.Query{
+					Type:   "directconnect-location",
+					Method: sdp.QueryMethod_GET,
+					Query:  *connection.Location,
+					Scope:  scope,
+				},
+				BlastPropagation: &sdp.BlastPropagation{
+					// Changes to the location will affect this, i.e., its speed, provider, etc.
+					In: true,
+					// We can't affect the location
+					Out: false,
+				},
+			})
+		}
+
+		if connection.LoaIssueTime != nil {
+			// +overmind:link directconnect-loa
+			item.LinkedItemQueries = append(item.LinkedItemQueries, &sdp.LinkedItemQuery{
+				Query: &sdp.Query{
+					Type:   "directconnect-loa",
+					Method: sdp.QueryMethod_GET,
+					Query:  *connection.ConnectionId,
+					Scope:  scope,
+				},
+				BlastPropagation: &sdp.BlastPropagation{
+					// Changes to the loa will affect this
+					In: true,
+					// We can't affect the loa
+					Out: false,
+				},
+			})
+		}
+
+		// Virtual Interfaces
+		item.LinkedItemQueries = append(item.LinkedItemQueries, &sdp.LinkedItemQuery{
+			// +overmind:link directconnect-virtual-interface
+			Query: &sdp.Query{
+				Type:   "directconnect-virtual-interface",
+				Method: sdp.QueryMethod_LIST,
```

Comment:
We can only use the `LIST` command here due to 1-n relation.
Is it OK to use it here?

Reply:
Not without applying some kind of filter here. `LinkedItemQueries` fully define the relationship here, so adding a LIST query would erroneously link this connection with ALL virtual-interfaces. 

Looking into the `DescribeVirtualInterfaces` API reference, that call can filter by `connectionId`, so a custom SEARCH query syntax like `connection:<id>` (or just the connection ARN if available) would be the way to go. This then would have to be implemented on the `directconnect-virtual-interface` source side to fill in the correct arguments to `DescribeVirtualInterfaces`.

Reply:
Thanks, I thought that the `Query:  *connection.ConnectionId` would do the trick, but after checking the `List` method implementation on the `DescribeOnlySource`, the query parameter will be ignored.
Is that right?

Reply:
LIST is "give me everything", ignoring the `Query`; SEARCH is "lookup a thing by this random `Query` string"; (and for completeness) GET is "give me the thing with this unique attribute value (also provided in `Query`".

Since search requests can get fancy with their queries, `DescribeOnlySource` has the `InputMapperSearch` func that takes a query string and transforms that into an `Input`. Most implementations of that we currently have only take a plain ARN. This case might be a bit more complicated.

Taking just the `ConnectionId` and returning

```go
DescribeVirtualInterfacesInput{
  ConnectionId: &query
}
```

is a decent start.

Since this is also the way customers would search in the UI and looking at the `DescribeVirtualInterfacesInput` struct, it would be great :tm: if there was a way to also search by `VirtualInterfaceId`. Unless there's an easy/stable way to distinguish `VirtualInterfaceId` and `ConnectionId` by their values, the next best solution is changing the `InputMapperSearch` interface to return `[]Input` and fixup https://github.com/overmindtech/aws-source/blob/d479ed14a49fc25db258f2cc6cd496b214f00983/sources/describe_source.go#L323 to iterate over the results (as well as the 10 or so `InputMapperSearch` implementations to return a single-item `[]Input` result).

Does that make sense? Happy to answer any questions.

Reply:
Thanks for the detailed answer @DavidS-ovm.
I'll go with the `SEARCH` method here and the query is `connection_id:<id>` as you suggested.
I'll keep these notes in mind while working on the `VirtualInterface` source.